plume-common/src/activity_pub/inbox.rs

```diff
@@ -1,48 +1,582 @@
+///    .with::<User, Announce, Message>()
+///    .with::<User, Create,   Message>()
+///    .done();
+/// ```
+pub enum Inbox<'a, C, E, R>
+where
+    E: From<InboxError<E>> + Debug,
+{
+    /// The activity has not been handled yet
+    ///
+    /// # Structure
+    ///
+    /// - the context to be passed to each handler.
+    /// - the activity
+    /// - the reason it has not been handled yet
+    NotHandled(&'a C, serde_json::Value, InboxError<E>),
 
-#[derive(Fail, Debug)]
-pub enum InboxError {
-    #[fail(display = "The `type` property is required, but was not present")]
-    NoType,
-    #[fail(display = "Invalid activity type")]
-    InvalidType,
-    #[fail(display = "Couldn't undo activity")]
-    CantUndo,
+    /// A matching handler have been found but failed
+    ///
+    /// The wrapped value is the error returned by the handler
+    Failed(E),
+
+    /// The activity was successfully handled
+    ///
+    /// The wrapped value is the value returned by the handler
+    Handled(R),
 }
 
-pub trait FromActivity<T: Object, C>: Sized {
-    type Error: From<ApError>;
+/// Possible reasons of inbox failure
+#[derive(Debug)]
+pub enum InboxError<E: Debug> {
+    /// None of the registered handlers matched
+    NoMatch,
 
-    fn from_activity(conn: &C, obj: T, actor: Id) -> Result<Self, Self::Error>;
+    /// No ID was provided for the incoming activity, or it was not a string
+    InvalidID,
 
-    fn try_from_activity(conn: &C, act: Create) -> Result<Self, Self::Error> {
-        Self::from_activity(
-            conn,
-            act.create_props.object_object()?,
-            act.create_props.actor_link::<Id>()?,
-        )
-    }
+    /// The activity type matched for at least one handler, but then the actor was
+    /// not of the expected type
+    InvalidActor(Option<E>),
+
+    /// Activity and Actor types matched, but not the Object
+    InvalidObject(Option<E>),
+
+    /// Error while dereferencing the object
+    DerefError,
 }
 
-pub trait Notify<C> {
-    type Error;
+impl<T: Debug> From<InboxError<T>> for () {
+    fn from(_: InboxError<T>) {}
+}
+
+/*
+ Type arguments:
+ - C: Context
+ - E: Error
+ - R: Result
+*/
+impl<'a, C, E, R> Inbox<'a, C, E, R>
+where
+    E: From<InboxError<E>> + Debug,
+{
+    /// Creates a new `Inbox` to handle an incoming activity.
+    ///
+    /// # Parameters
+    ///
+    /// - `ctx`: the context to pass to each handler
+    /// - `json`: the JSON representation of the incoming activity
+    pub fn handle(ctx: &'a C, json: serde_json::Value) -> Inbox<'a, C, E, R> {
+        Inbox::NotHandled(ctx, json, InboxError::NoMatch)
+    }
+
+    /// Registers an handler on this Inbox.
+    pub fn with<A, V, M>(self) -> Inbox<'a, C, E, R>
+    where
+        A: AsActor<&'a C> + FromId<C, Error = E>,
+        V: activitypub::Activity,
+        M: AsObject<A, V, &'a C, Error = E> + FromId<C, Error = E>,
+        M::Output: Into<R>,
+    {
+        if let Inbox::NotHandled(ctx, act, e) = self {
```

Comment:
is there such a thing as `if not let`? so we can move this whole block out by one indent, by returning early?

Reply:
I don't think so…

Reply:
There is match